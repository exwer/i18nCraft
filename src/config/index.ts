export * from './validator'
export * from './defaults'
export * from './types'

import { validateConfig } from './validator'
import { getDefaultConfig } from './defaults'
import type { CLIConfig, ValidationResult } from './types'

// 重新导出函数
export { validateConfig, getDefaultConfig }

/**
 * 配置管理器
 */
export class ConfigManager {
  private config: CLIConfig

  constructor(config?: Partial<CLIConfig>) {
    this.config = { ...getDefaultConfig(), ...config }
  }

  /**
   * 获取配置
   */
  getConfig(): CLIConfig {
    return this.config
  }

  /**
   * 更新配置
   */
  updateConfig(updates: Partial<CLIConfig>): void {
    this.config = { ...this.config, ...updates }
  }

  /**
   * 验证配置
   */
  validate(): ValidationResult {
    return validateConfig(this.config)
  }

  /**
   * 获取有效配置（如果无效则抛出错误）
   */
  getValidConfig(): CLIConfig {
    const validation = this.validate()
    if (!validation.valid) {
      throw new Error(`Invalid configuration: ${validation.errors.join(', ')}`)
    }
    return this.config
  }
} 
