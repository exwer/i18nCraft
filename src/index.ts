// 核心类型导出
export type { 
  TransformOptions, 
  LocaleConfig,
  LocaleObject,
  CLIConfig,
  TextMatch,
  TransformResult
} from './types'

// 核心转换功能
export { transformSFC } from './core/transform'

// 新的转换器架构
export { 
  BaseTransformer, 
  VueTransformer, 
  ReactTransformer,
  JavaScriptTransformer,
  createTransformer,
  transform
} from './core/transformer'

// Parser 系统
export { RecastParser } from './core/parsers'
export type { RecastParseOptions, RecastTransformOptions } from './core/parsers'

// Provider 系统
export type { I18nProvider } from './core/providers'
export { 
  ReactI18nextProvider,
  VueI18nProvider,
  VanillaI18nProvider,
  VanillaFunctionProvider,
  VanillaI18nextProvider,
  createVanillaProvider
} from './core/providers'

// 错误处理
export { 
  I18nCraftError, 
  ErrorCode, 
  createError
} from './core/errors'

// 文本匹配器
export { TextMatcher } from './core/matcher'

// React转换器
export { transformReact } from './plugins/react'

// 构建工具插件
export { createUnplugin } from './plugins/unplugin'
export { createVitePlugin } from './plugins/vite'
export { createWebpackPlugin } from './plugins/webpack'
