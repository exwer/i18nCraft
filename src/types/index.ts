// 基础类型定义
export interface LocaleObject {
  [key: string]: string | LocaleObject
}

export interface LocaleConfig {
  [language: string]: LocaleObject
}

// 转换格式类型
export type TransformFormatFunction = (key: string) => string
export type TransformFormatString = string

export interface TransformFormat {
  template: TransformFormatFunction | TransformFormatString
  script: TransformFormatFunction | TransformFormatString
  interpolation: TransformFormatFunction | TransformFormatString
}

// 转换选项类型
export interface TransformOptions {
  locale: LocaleConfig
  enableScript?: boolean
  enableTemplate?: boolean
  customMatcher?: (text: string) => string | false
  keyGenerator?: (text: string) => string
  debug?: boolean
  transformFormat?: TransformFormat
}

// CLI 配置类型
export interface CLIConfig {
  scanDir: string
  outDir: string
  exts: string[]
  locale: LocaleConfig
  transformFormat?: TransformFormat
  enableScript?: boolean
  enableTemplate?: boolean
  debug?: boolean
}

// 匹配结果类型
export type MatchResult = string | false

// 文本匹配类型
export interface TextMatch {
  original: string
  key: string
  type: string
  line?: number
  column?: number
}

// 转换结果类型
export interface TransformResult {
  code: string
  matches: TextMatch[]
  errors: Error[]
}

// 插件选项类型
export interface PluginOptions {
  locale: LocaleConfig
  transformFormat?: TransformFormat
}

// 文件处理结果类型
export interface FileProcessResult {
  success: boolean
  filePath: string
  outputPath?: string
  error?: string
  stats?: {
    transformedStrings: number
    processingTime: number
  }
}

// 批量处理结果类型
export interface BatchProcessResult {
  totalFiles: number
  successfulFiles: number
  failedFiles: number
  results: FileProcessResult[]
  totalTime: number
}

// AST 节点类型（用于 recast）
export interface ASTNode {
  type: string
  [key: string]: any
}

// 路径类型（用于 recast）
export interface ASTPath {
  node: ASTNode
  parent: ASTPath | null
  [key: string]: any
}

// 错误类型增强
export interface ErrorContext {
  filePath?: string
  line?: number
  column?: number
  code?: string
  [key: string]: any
}

// 验证结果类型
export interface ValidationResult {
  valid: boolean
  errors: string[]
  warnings: string[]
}

// 性能指标类型
export interface PerformanceMetrics {
  parseTime: number
  transformTime: number
  totalTime: number
  memoryUsage: number
}

// 导出所有类型
export * from '../core/errors' 
