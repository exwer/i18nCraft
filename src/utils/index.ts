import type babelCore from '@babel/core'
import type { LocaleMsg } from '../index'

export function hasImportedMember(
  node: babelCore.types.ImportDeclaration,
  name: string,
) {
  return node.specifiers.some(
    item =>
      item.type === 'ImportSpecifier'
      && item.imported.type === 'Identifier'
      && item.imported.name === name,
  )
}

export function getMatchedMsgPath(
  locale: LocaleMsg,
  target: string,
): false | string {
  const lang = locale[Object.keys(locale)?.[0]]
  if (!lang)
    return false
  function getPath(
    obj: Record<string, string | Record<string, any>>,
    value: string,
  ) {
    let result = ''
    const keys = Object.keys(obj)
    for (let i = 0; i < keys.length; i++) {
      const key = keys[i]
      if (obj[key] === value)
        return `.${key}`
      else if (typeof obj[key] !== 'string')
        result += getPath(obj[key] as any, value)
    }
    result = ''

    return result
  }
  const result = getPath(lang, target)
  if (result)
    return result
  return false
}
