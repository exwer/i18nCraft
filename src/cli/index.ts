import path from 'path'
import fs from 'fs'
import { ConfigManager } from '../config'
import { transformSFC } from '../core/transform'
import { globalMiddlewareManager } from '../core/middleware'
import { getBuiltinMiddlewares } from '../core/middleware/builtin'
import type { CLIConfig } from '../config/types'

/**
 * 新的CLI工具类
 */
export class I18nCraftCLI {
  private configManager: ConfigManager
  private initialized = false

  constructor(config?: Partial<CLIConfig>) {
    this.configManager = new ConfigManager(config)
  }

  /**
   * 初始化CLI
   */
  async init(): Promise<void> {
    if (this.initialized) return

    // 注册内置中间件
    const builtinMiddlewares = getBuiltinMiddlewares()
    builtinMiddlewares.forEach((middleware: any) => {
      globalMiddlewareManager.register(middleware)
    })

    this.initialized = true
  }

  /**
   * 扫描文件
   */
  scanFiles(scanDir: string, exts: string[]): string[] {
    const files: string[] = []
    
    function scanDirectory(dir: string): void {
      if (!fs.existsSync(dir)) return
      
      const items = fs.readdirSync(dir)
      for (const item of items) {
        const fullPath = path.join(dir, item)
        const stat = fs.statSync(fullPath)
        
        if (stat.isDirectory()) {
          scanDirectory(fullPath)
        } else if (stat.isFile()) {
          const ext = path.extname(fullPath)
          if (exts.includes(ext)) {
            files.push(fullPath)
          }
        }
      }
    }
    
    scanDirectory(scanDir)
    return files
  }

  /**
   * 转换单个文件
   */
  async transformFile(filePath: string, config: CLIConfig): Promise<{
    success: boolean
    outputPath?: string
    error?: string
    stats?: {
      transformedStrings: number
      processingTime: number
    }
  }> {
    const startTime = Date.now()
    
    try {
      const source = fs.readFileSync(filePath, 'utf-8')
      const result = await transformSFC(source, {
        locale: config.locale,
        enableScript: config.enableScript,
        enableTemplate: config.enableTemplate,
        customMatcher: config.customMatcher,
        keyGenerator: config.keyGenerator,
        debug: config.debug,
        transformFormat: config.transformFormat
      })

      // 计算输出路径
      const relPath = path.relative(config.scanDir, filePath)
      const outputPath = path.join(config.outDir, relPath)
      const outputDir = path.dirname(outputPath)
      
      // 确保输出目录存在
      fs.mkdirSync(outputDir, { recursive: true })
      
      // 写入文件
      fs.writeFileSync(outputPath, result, 'utf-8')
      
      const processingTime = Date.now() - startTime
      
      return {
        success: true,
        outputPath,
        stats: {
          transformedStrings: 0, // 这里可以从result中获取
          processingTime
        }
      }
    } catch (error) {
      return {
        success: false,
        error: error instanceof Error ? error.message : String(error)
      }
    }
  }

  /**
   * 批量转换文件
   */
  async batchTransform(config?: Partial<CLIConfig>): Promise<{
    totalFiles: number
    successfulFiles: number
    failedFiles: number
    totalTime: number
    errors: string[]
  }> {
    await this.init()
    
    if (config) {
      this.configManager.updateConfig(config)
    }
    
    const finalConfig = this.configManager.getValidConfig()
    const files = this.scanFiles(finalConfig.scanDir, finalConfig.exts)
    
    console.log(`[i18ncraft] Found ${files.length} files to process`)
    
    const startTime = Date.now()
    let successfulFiles = 0
    let failedFiles = 0
    const errors: string[] = []
    
    for (const file of files) {
      const result = await this.transformFile(file, finalConfig)
      
      if (result.success) {
        successfulFiles++
        console.log(`[i18ncraft] ✓ ${file} -> ${result.outputPath}`)
      } else {
        failedFiles++
        errors.push(`${file}: ${result.error}`)
        console.error(`[i18ncraft] ✗ ${file}: ${result.error}`)
      }
    }
    
    const totalTime = Date.now() - startTime
    
    console.log(`[i18ncraft] Batch transformation completed`)
    console.log(`[i18ncraft] Total: ${files.length}, Success: ${successfulFiles}, Failed: ${failedFiles}`)
    console.log(`[i18ncraft] Total time: ${totalTime}ms`)
    
    return {
      totalFiles: files.length,
      successfulFiles,
      failedFiles,
      totalTime,
      errors
    }
  }

  /**
   * 获取配置
   */
  getConfig(): CLIConfig {
    return this.configManager.getConfig()
  }

  /**
   * 更新配置
   */
  updateConfig(config: Partial<CLIConfig>): void {
    this.configManager.updateConfig(config)
  }

  /**
   * 验证配置
   */
  validateConfig(): { valid: boolean; errors: string[]; warnings: string[] } {
    return this.configManager.validate()
  }
} 
