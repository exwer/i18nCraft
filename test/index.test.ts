import * as fs from 'fs'
import * as path from 'path'
import { describe, expect, it } from 'vitest'
import * as VueCompiler from '@vue/compiler-sfc'
import { start } from '../src/index'
function compile(path: string, target: string) {
  try {
    let result = ''
    const source = fs.readFileSync(
      path,
      'utf-8',
    )
    const { descriptor } = VueCompiler.parse(source)
    const { content: script } = VueCompiler.compileScript(descriptor, {
      id: '123',
    })
    const { code: template } = VueCompiler.compileTemplate({
      source: descriptor.template!.content.toString(),
      filename: 'fuck',
      id: '456',
    })
    result += script
    result += template
    fs.writeFileSync(target, result)
  }
  catch (error) {
    console.error(error)
  }
}
describe('import test', () => {
  it('single script', async() => {
    const source = '<script>import { a } from "b"</script>'
    const result = await start(source)
    expect(result).toMatchSnapshot()
  })
  it('complete vue file', async() => {
    const source = '<template><div></div></template><script></script><style></style>'
    const result = await start(source)
    expect(result).toMatchSnapshot()
  })
  it('repeat import', async() => {
    const source = '<script>import {ref} from "vue";import {useI18n} from "vue-i18n"</script>'
    const result = await start(source)
    expect(result).toMatchSnapshot()
  })
})
