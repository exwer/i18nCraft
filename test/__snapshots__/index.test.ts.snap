// Vitest Snapshot v1

exports[`import test > complete vue file 1`] = `"<template><div></div></template><script></script><style></style>"`;

exports[`import test > repeat import 1`] = `
"<script>import { ref, ref, computed } from \\"vue\\";
import { useI18n, useI18n } from \\"vue-i18n\\";</script>"
`;

exports[`import test > single script 1`] = `
"<script>import { ref } from \\"vue\\";
import { useI18n } from \\"vue-i18n\\";
import { a } from \\"b\\";</script>"
`;
